### PR TITLE
Move AccumuloClient.getInstanceID to InstanceOperations

### PR DESCRIPTION
**Pull request by jzgithub1**
Made change that was requested.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java [hunk 1]
```diff
@@ -130,4 +130,11 @@ boolean testClassLoad(final String className, final String asTypeName)
    * @since 1.7.0
    */
   void waitForBalance() throws AccumuloException;
+
+  /**
+   * Returns a unique string that identifies this instance of accumulo.
+   *
+   * @return a String
+   */
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java]: This needs a `@since 2.0.0` java doc tag

**Comment by keith-turner** [comment]
Also need to remove getInstanceId method from AccumuloClient

File: core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java [hunk 2]
```diff
@@ -232,4 +232,10 @@ public static String lookupInstanceName(ZooCache zooCache, UUID instanceId) {
     }
     return null;
   }
+
+  @Override
+  public String getInstanceID() {
```
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java]: The `getInstanceID` method still needs to be removed from the `AccumuloClient` interface but leave it in `ClientContext`

Comment (mikewalch) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java]: Just convert all uses of `AccumuloClient.getInstanceID` to `AccumuloClient.instanceOperations().getInstanceID`.  Any calls to `ServerContext.getInstanceID()` or `ClientContext.getIntanceID` should not need to be changed.  If you need to mock out the call to `AccumuloClient.instanceOperations().getInstanceID` do the following: 

```
--- a/server/master/src/test/java/org/apache/accumulo/master/replication/SequentialWorkAssignerTest.java
+++ b/server/master/src/test/java/org/apache/accumulo/master/replication/SequentialWorkAssignerTest.java
@@ -26,6 +26,7 @@ import java.util.Map;
 import java.util.TreeMap;
 
 import org.apache.accumulo.core.client.AccumuloClient;
+import org.apache.accumulo.core.client.admin.InstanceOperations;
 import org.apache.accumulo.core.clientImpl.Table;
 import org.apache.accumulo.core.conf.AccumuloConfiguration;
 import org.apache.accumulo.core.replication.ReplicationConstants;
@@ -71,7 +72,9 @@ public class SequentialWorkAssignerTest {
     assigner.setWorkQueue(workQueue);
     assigner.setQueuedWork(queuedWork);
 
-    expect(client.getInstanceID()).andReturn("instance");
+    InstanceOperations opts = createMock(InstanceOperations.class);
+    expect(opts.getInstanceID()).andReturn("instance");
+    expect(client.instanceOperations()).andReturn(opts);
 
     // file1 replicated
     expect(zooCache.get(ZooUtil.getRoot("instance") + ReplicationConstants.ZOO_WORK_QUEUE + "/"
@@ -85,7 +88,7 @@ public class SequentialWorkAssignerTest {
                     new ReplicationTarget("cluster1", "2", Table.ID.of("2")))))
                         .andReturn(new byte[0]);
 
-    replay(workQueue, zooCache, client);
+    replay(workQueue, zooCache, opts, client);
 
     assigner.cleanupFinishedWork();
```

**Comment by jzgithub1** [comment]
Made requested code changes.  Thanks for all the help.

File: core/src/main/java/org/apache/accumulo/core/client/ScannerBase.java [hunk 3]
```diff
@@ -305,5 +305,7 @@
    *
    * @since 2.0.0
    */
-  void setExecutionHints(Map<String,String> hints);
+  default void setExecutionHints(Map<String,String> hints) {
```
Comment (mikewalch) [inline core/src/main/java/org/apache/accumulo/core/client/ScannerBase.java]: It looks like there are some unrelated changes in this PR

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeChooserEnvironment.java [hunk 4]
```diff
@@ -35,17 +36,33 @@
...
+  public VolumeChooserEnvironment(Table.ID tableId, Text endRow, ServerContext context) {
     this.context = context;
     this.scope = ChooserScope.TABLE;
     this.tableId = Objects.requireNonNull(tableId);
+    this.endRow = endRow;
+
+  }
+
+  /**
+   * The end row of the tablet for which a volume is being chosen. Only call this when the scope is
```
Comment (mikewalch) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeChooserEnvironment.java]: More unrelated changes

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java [hunk 5]
```diff
@@ -914,7 +914,8 @@ public static void cloneTable(ServerContext context, Table.ID srcTableId, Table.
         Key k = entry.getKey();
         Mutation m = new Mutation(k.getRow());
         m.putDelete(k.getColumnFamily(), k.getColumnQualifier());
-        VolumeChooserEnvironment chooserEnv = new VolumeChooserEnvironment(tableId, context);
+        VolumeChooserEnvironment chooserEnv = new VolumeChooserEnvironment(tableId,
```
Comment (mikewalch) [inline server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java]: More unrelated changes

File: server/master/src/test/java/org/apache/accumulo/master/replication/UnorderedWorkAssignerTest.java [hunk 6]
```diff
@@ -24,15 +24,11 @@
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 
-import java.util.Arrays;
-import java.util.HashSet;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (mikewalch) [inline server/master/src/test/java/org/apache/accumulo/master/replication/UnorderedWorkAssignerTest.java]: This is causing checkstyle to fail.  Your IDE might have done this for you but we don't use start imports.
```
src/test/java/org/apache/accumulo/master/replication/UnorderedWorkAssignerTest.java:[27] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - java.util.*.
```

**Comment by jzgithub1** [comment]
I merged with what was in the upstream master branch.  That is how those unrelated changes got in.  I don't recall editing those unrelated  files.

**Comment by keith-turner** [comment]
> I merged with what was in the upstream master branch.

Personally, I try to avoid rebasing or merging master into a branch/PR that is under review because I think it makes it harder for the reviewer to follow whats going on.  Sometimes its necessary though.  

In this case I am mystified as to what is happening because I thought changes that were already made in master would not show up in the diff.

**Comment by mikewalch** [comment]
>  I don't recall editing those unrelated files.

I am not sure how it happened but the files were changed in a commit (e683dd0dbf8e48099b9f7f0e8ad8a80e35a06d98) of yours for this PR.

**Comment by jzgithub1** [comment]
Since I had been working on this a while I though it would be good to pull the upstream master and merge to the branch so I could work around any changes that someone was making to the files I was working on.  Sorry about the confusion.

**Comment by jzgithub1** [comment]
I did some rebasing on my branch.  I must have squashed a commit that occurred on upstream/master.

**Comment by keith-turner** [comment]
> Since I had been working on this a while I though it would be good to pull the upstream master and merge to the branch so I could work around any changes that someone was making to the files I was working on.

Makes sense.  When I am reviewing I am more than happy to resolve conflicts when merging.

**Comment by ctubbsii** [comment]
> Since I had been working on this a while I though it would be good to pull the upstream master and merge to the branch so I could work around any changes that someone was making to the files I was working on. Sorry about the confusion.

In general, if you have something long-lived, it's better to rebase. Merging your the branch that contained your merge-base frequently creates confusion like this. Also, even if you did successfully merge without these kinds of issues, if somebody rebases again later, they'll probably have to re-resolve merge conflicts. So, it doesn't save any work to merge in. Rebase, on the other hand, keeps all the commit history linear in your feature branch, and is less likely to have these kinds of merge artifacts.

I frequently do interactive rebase with long-lived branches (`git rebase -i`)

**Comment by jzgithub1** [comment]
Would you like me to just re-branch from master and redo all the work?  Its really no problem for me to do that if it will preserve history correctly.

**Comment by ctubbsii** [comment]
You can just rebase, fix any conflicts, and do a force push, and it will update this PR. If it's easier to reset this branch to the master commit, and add a new commit, and force push that, that'll work, too (effectively the same thing).

**Comment by jzgithub1** [comment]
Just pushed the branch again.  Let me know if there are any problems.  If there are I will just start over from scratch.  No big deal.

**Comment by mikewalch** [comment]
I think you still need to rebase with master.  Follow something like below:
```
git checkout master
git fetch upstream
git pull upstream master
git checkout accumulo-907_
git rebase master
git push -f origin accumulo-907_
```

**Comment by ctubbsii** [comment]
For future reference, the following instructions for rebase are even more succinct:

```bash
git remote update # optionally add --prune to update deleted branches
git checkout accumulo-907_ # if you're not already in that branch locally
git rebase upstream/master # the actual rebase, I recommend doing interactive with -i
git push -f origin accumulo-907_ # to update your fork and the PR
```
